File: src/irp/runtime/forwarding.py
from __future__ import annotations

import logging
import subprocess
from typing import Dict, Iterable

from irp.model.routing import ForwardingEntry
from irp.runtime.config import ForwardingConfig


class ForwardingApplier:
    def apply(self, entries: Iterable[ForwardingEntry]) -> None:
        raise NotImplementedError


class NullForwardingApplier(ForwardingApplier):
    def apply(self, entries: Iterable[ForwardingEntry]) -> None:
        _ = list(entries)


class LinuxForwardingApplier(ForwardingApplier):
    def __init__(self, cfg: ForwardingConfig, logger: logging.Logger) -> None:
        self._cfg = cfg
        self._log = logger
        self._installed: Dict[int, ForwardingEntry] = {}

    def apply(self, entries: Iterable[ForwardingEntry]) -> None:
        desired: Dict[int, ForwardingEntry] = {entry.destination: entry for entry in entries}
        stale_destinations = sorted(set(self._installed.keys()) - set(desired.keys()))
        for destination in stale_destinations:
            stale = self._installed[destination]
            self._delete_route(stale)
            self._installed.pop(destination, None)

        for destination, entry in sorted(desired.items()):
            current = self._installed.get(destination)
            if current == entry:
                continue
            self._replace_route(entry)
            self._installed[destination] = entry

    def _replace_route(self, entry: ForwardingEntry) -> None:
        prefix = self._cfg.destination_prefixes.get(entry.destination)
        via = self._cfg.next_hop_ips.get(entry.next_hop)
        if prefix is None or via is None:
            self._log.warning(
                "Skip FIB install for dst=%s next_hop=%s: missing prefix/via mapping",
                entry.destination,
                entry.next_hop,
            )
            return
        cmd = ["ip", "route", "replace", prefix, "via", via, "table", str(self._cfg.table)]
        self._run_cmd(cmd)

    def _delete_route(self, entry: ForwardingEntry) -> None:
        prefix = self._cfg.destination_prefixes.get(entry.destination)
        if prefix is None:
            return
        cmd = ["ip", "route", "del", prefix, "table", str(self._cfg.table)]
        self._run_cmd(cmd, ignore_error=True)

    def _run_cmd(self, cmd: list[str], ignore_error: bool = False) -> None:
        if self._cfg.dry_run:
            self._log.info("FIB dry-run: %s", " ".join(cmd))
            return
        try:
            subprocess.run(cmd, check=True, capture_output=True, text=True)
        except subprocess.CalledProcessError as exc:
            if ignore_error:
                self._log.debug("FIB command ignored failure: %s -> %s", " ".join(cmd), exc.stderr)
                return
            raise RuntimeError(f"FIB command failed: {' '.join(cmd)}\n{exc.stderr}") from exc

File: src/irp_rust/src/runtime/forwarding.rs
use std::collections::BTreeMap;
use std::process::Command;

use anyhow::{Context, Result};
use tracing::{debug, info, warn};

use crate::model::routing::ForwardingEntry;
use crate::runtime::config::ForwardingConfig;

pub trait ForwardingApplier: Send {
    fn apply(&mut self, entries: &[ForwardingEntry]) -> Result<()>;
}

pub struct NullForwardingApplier;

impl ForwardingApplier for NullForwardingApplier {
    fn apply(&mut self, _entries: &[ForwardingEntry]) -> Result<()> {
        Ok(())
    }
}

pub struct LinuxForwardingApplier {
    cfg: ForwardingConfig,
    installed: BTreeMap<u32, ForwardingEntry>,
}

impl LinuxForwardingApplier {
    pub fn new(cfg: ForwardingConfig) -> Self {
        Self {
            cfg,
            installed: BTreeMap::new(),
        }
    }

    fn replace_route(&self, entry: &ForwardingEntry) -> Result<()> {
        let Some(prefix) = self.cfg.destination_prefixes.get(&entry.destination) else {
            warn!(
                "skip FIB install for dst={} next_hop={}: missing destination_prefixes mapping",
                entry.destination, entry.next_hop
            );
            return Ok(());
        };
        let Some(via) = self.cfg.next_hop_ips.get(&entry.next_hop) else {
            warn!(
                "skip FIB install for dst={} next_hop={}: missing next_hop_ips mapping",
                entry.destination, entry.next_hop
            );
            return Ok(());
        };

        let cmd = [
            "ip".to_string(),
            "route".to_string(),
            "replace".to_string(),
            prefix.to_string(),
            "via".to_string(),
            via.to_string(),
            "table".to_string(),
            self.cfg.table.to_string(),
        ];
        self.run_cmd(&cmd, false)
    }

    fn delete_route(&self, entry: &ForwardingEntry) -> Result<()> {
        let Some(prefix) = self.cfg.destination_prefixes.get(&entry.destination) else {
            return Ok(());
        };
        let cmd = [
            "ip".to_string(),
            "route".to_string(),
            "del".to_string(),
            prefix.to_string(),
            "table".to_string(),
            self.cfg.table.to_string(),
        ];
        self.run_cmd(&cmd, true)
    }

    fn run_cmd(&self, cmd: &[String], ignore_error: bool) -> Result<()> {
        if self.cfg.dry_run {
            info!("FIB dry-run: {}", cmd.join(" "));
            return Ok(());
        }

        let output = Command::new(&cmd[0])
            .args(&cmd[1..])
            .output()
            .with_context(|| format!("failed to execute command: {}", cmd.join(" ")))?;

        if output.status.success() {
            return Ok(());
        }

        if ignore_error {
            debug!(
                "ignore FIB cmd failure: {} -> {}",
                cmd.join(" "),
                String::from_utf8_lossy(&output.stderr)
            );
            return Ok(());
        }

        anyhow::bail!(
            "FIB command failed: {}\n{}",
            cmd.join(" "),
            String::from_utf8_lossy(&output.stderr)
        )
    }
}

impl ForwardingApplier for LinuxForwardingApplier {
    fn apply(&mut self, entries: &[ForwardingEntry]) -> Result<()> {
        let desired: BTreeMap<u32, ForwardingEntry> = entries
            .iter()
            .map(|entry| (entry.destination, entry.clone()))
            .collect();

        let stale_destinations: Vec<u32> = self
            .installed
            .keys()
            .filter(|destination| !desired.contains_key(destination))
            .copied()
            .collect();

        for destination in stale_destinations {
            if let Some(stale) = self.installed.get(&destination) {
                self.delete_route(stale)?;
            }
            self.installed.remove(&destination);
        }

        for (destination, entry) in &desired {
            let current = self.installed.get(destination);
            if current == Some(entry) {
                continue;
            }
            self.replace_route(entry)?;
            self.installed.insert(*destination, entry.clone());
        }

        Ok(())
    }
}
